docs/content/guides/dagster-pipes/modifying-existing-code-to-work-with-dagster-pipes.mdx

```diff
@@ -1,6 +1,270 @@
+
+  </Note>
+
+- An instance of the Dagster Pipes context via <PyObject module="dagster_pipes" object="PipesContext" method="get"/>.
+
+In our sample Python script, the changes would look like the following:
+
+```python file=/guides/dagster/dagster_pipes/subprocess/external_code_with_context.py
+import pandas as pd
+from dagster_pipes import PipesContext, open_dagster_pipes
+
+
+def main():
+    orders_df = pd.DataFrame({"order_id": [1, 2], "item_id": [432, 878]})
+    total_orders = len(orders_df)
+    # get the Dagster Pipes context
+    context = PipesContext.get()
+    ...
+
+
+if __name__ == "__main__":
+    # connect to Dagster Pipes
+    with open_dagster_pipes() as context:
+        main()
+```
+
+---
+
+### Send log messages to Dagster
+
+Dagster Pipes context offers a built-in logging capability that enables you to stream log messages back to Dagster:
+
+```python file=/guides/dagster/dagster_pipes/subprocess/external_code_with_logs.py
+import pandas as pd
+from dagster_pipes import PipesContext, open_dagster_pipes
+
+
+def main():
+    orders_df = pd.DataFrame({"order_id": [1, 2], "item_id": [432, 878]})
+    total_orders = len(orders_df)
+    # get the Dagster Pipes context
+    context = PipesContext.get()
+    # send log messages back to Dagster
+    context.log.info(f"processing total {total_orders} orders")
+
+
+if __name__ == "__main__":
+    # connect to Dagster Pipes
+    with open_dagster_pipes() as context:
+        main()
+```
+
+- <PyObject module="dagster_pipes" object="open_dagster_pipes" /> constructs a context
+  manager that establishes a connection with Dagster Pipes. It returns a <PyObject
+  module="dagster_pipes"
+  object="PipesContext"
+  /> object.
+- Inside the context, `context.log.info(...)` is a method call that sends a log message back to Dagster.
+
+Then, the log messages will show up in the Dagster UI. You can filter the log levels to only view `info` level messages:
+
+1. Click the **Levels** filter next to the log filter field. This will present a dropdown of all log levels.
+2. Select the **info** checkbox and deselect the others. This will show only the logs marked as **info** level.
+
+<Image
+alt="Send log messages to Dagster"
+src="/images/guides/dagster-pipes/modifying-code-4-log-messages.png"
+width={1326}
+height={742}
+/>
+
+### Send structured metadata to Dagster
+
+Sometimes, you may want to log information from your external code to be structured metadata in Dagster. Dagster Pipes context also comes the ability to log structured metadata back to Dagster.
+
+#### Report asset materialization
+
+Similar to [reporting materialization metadata within the Dagster process](/concepts/assets/software-defined-assets#reporting-materialization-metadata), you can also report asset materialization back to Dagster from the external process.
```

Comment:
Are we expecting some baseline familiarity with dagster? I'm not sure how much this reference link helps if we think a lot of non dagster users will be looking at this page